fili-core/src/main/java/com/yahoo/bard/webservice/web/handlers/CacheWeightCheckRequestHandler.java

```diff
@@ -0,0 +1,225 @@
+                    context,
+                    request,
+                    druidQuery,
+                    weightCheckResponse,
+                    queryRowLimit,
+                    cacheKey,
+                    cacheService,
+                    weightCheckRequestHandler,
+                    true
+            );
+        }
+        HttpErrorCallback error = response.getErrorCallback(druidQuery);
+        FailureCallback failure = response.getFailureCallback(druidQuery);
+        webService.postDruidQuery(context, weightQuerySuccess, error, failure, weightEvaluationQuery);
+        return true;
+    }
+
+    /**
+     * Build a callback that writes to the cache if request wasn't already cached. It then delegates to the
+     * WeightCheckRequestHandler callback which continues the original request or refuses it with an
+     * HTTP INSUFFICIENT_STORAGE (507)status based on the cardinality of the requester 's query as
+     * measured by the weight check query.
+     *
+     * @param context  The context data from the request processing chain
+     * @param request  The API request itself
+     * @param druidQuery  The query being processed
+     * @param response  The response handler
+     * @param queryRowLimit  The number of aggregating lines allowed
+     * @param cacheKey Key into which to write a cache entry
+     * @param cacheService CacheService object
+     * @param weightCheckRequestHandler WeightCheckRequestHandler object
+     * @param writeCache Read or Write cache flag
+     *
+     * @return The callback handler for the weight request
+     */
+    protected SuccessCallback buildCacheSuccessCallback(
+            final RequestContext context,
+            final DataApiRequest request,
+            final DruidAggregationQuery<?> druidQuery,
+            final ResponseProcessor response,
+            final long queryRowLimit,
+            final String cacheKey,
+            final CacheService cacheService,
+            final WeightCheckRequestHandler weightCheckRequestHandler,
+            final Boolean writeCache
+    ) {
+        return new SuccessCallback() {
+            SuccessCallback superSuccessCallback = weightCheckRequestHandler.buildSuccessCallback(
+                    context,
+                    request,
+                    druidQuery,
+                    response,
+                    queryRowLimit
+            );
+            @Override
+            public void invoke(JsonNode jsonResult) {
+                if (writeCache) {
+                    cacheService.writeCache(
+                            response,
+                            jsonResult,
+                            dataCache,
+                            druidQuery,
+                            querySigningService,
+                            cacheKey,
+                            writer
+                    );
+                }
+                superSuccessCallback.invoke(jsonResult);
+            }
+        };
+    }
+
```

Comment:
This can be highly simplified if we simply pass the delegate into the method after building it in the caller.

```suggestion
    /**
     * Build a callback that writes to the cache if request wasn't already cached. It then delegates to the
     * WeightCheckRequestHandler callback which continues the original request or refuses it with an
     * HTTP INSUFFICIENT_STORAGE (507)status based on the cardinality of the requester 's query as
     * measured by the weight check query.
     *
     * @return The callback handler for the weight request
     */
    protected SuccessCallback buildCacheSuccessCallback(
            SuccessCallback successCallback,
            CacheService cacheService,
            DruidAggregationQuery druidQuery,
            ResponseProcessor response
    ) {
        return new SuccessCallback() {
            @Override
            public void invoke(JsonNode jsonResult) {
                // send the response to the user before waiting to cache
                successCallback.invoke(jsonResult);
                try {
                    cacheService.writeCache(response, jsonResult, druidQuery);
                } catch (JsonProcessingException e) {
                    // Warn on cache write exception only
                }
            }
        };
    };

```